src/sage/graphs/graph_decompositions/tree_decomposition.pyx

```diff
@@ -761,18 +781,275 @@ def treewidth(g, k=None, kmin=None, certificate=False, algorithm=None):
+                directed_tree.add_path([current_node, new_node, neighbor])
+                bag[new_node] = bag[current_node]
+
+    # Step 5: If the node v has only one child, then it is either an introduce
+    # node or a forget node.
+    def add_path_of_intro_nodes(u, v):
+        """
+        Replace the arc (u, v) by a path of introduce nodes.
+        """
+        if len(bag[u]) + 1 == len(bag[v]):
+            return
+
+        diff = list(bag[v] - bag[u])
+        diff.pop()
+
+        last_node = u
+        for w in diff:
+            new_node = directed_tree.add_vertex()
+            bag[new_node] = bag[last_node].union(Set((w,)))
+            directed_tree.add_edge(last_node, new_node)
+            last_node = new_node
+
+        directed_tree.add_edge(last_node, v)
+        directed_tree.delete_edge(u, v)
+
+    def add_path_of_forget_nodes(u, v):
+        """
+        Replace the arc (u, v) by a path of forget nodes.
+        """
+        if len(bag[v]) + 1 == len(bag[u]):
+            return
+
+        diff = list(bag[u] - bag[v])
+        diff.pop()
+
+        last_node = u
+        for w in diff:
+            new_node = directed_tree.add_vertex()
+            bag[new_node] = bag[last_node] - {w}
+            directed_tree.add_edge(last_node, new_node)
+            last_node = new_node
+
+        directed_tree.add_edge(last_node, v)
+        directed_tree.delete_edge(u, v)
+
+    for ui in list(directed_tree):
+        if directed_tree.out_degree(ui) != 1:
+            continue
+
+        vi = next(directed_tree.neighbor_out_iterator(ui))
+        bag_ui, bag_vi = bag[ui], bag[vi]
+
+        # Merge the nodes if the two bags are the same
+        if bag_ui == bag_vi:
+            if directed_tree.in_degree(ui) == 1:
+                parent = next(directed_tree.neighbor_in_iterator(ui))
+                directed_tree.add_edge(parent, vi)
+            else:
+                root = vi
+            directed_tree.delete_vertex(ui)
+
+        # Add paths of intro / forget nodes accordingly
+
+        elif bag_ui.issubset(bag_vi):
+            add_path_of_intro_nodes(ui, vi)
+
+        elif bag_vi.issubset(bag_ui):
+            add_path_of_forget_nodes(ui, vi)
+
+        # Handle the case when the two nodes are not related in any way above
+        else:
+            wi = directed_tree.add_vertex()
+            bag[wi] = bag[ui] & bag[vi]
+            directed_tree.add_path([ui, wi, vi])
+            directed_tree.delete_edge(ui, vi)
+            add_path_of_forget_nodes(ui, wi)
+            add_path_of_intro_nodes(wi, vi)
+
+    # Return the nice tree decomposition after the processing
+    nice_tree_decomp = Graph(directed_tree, name=name)
+
+    bfs_ordering = nice_tree_decomp.breadth_first_search(start=root)
+    relabeling = {u: (i, bag[u]) for i, u in enumerate(bfs_ordering)}
+    nice_tree_decomp.relabel(inplace=True, perm=relabeling)
+
+    return nice_tree_decomp
+
+def label_nice_tree_decomposition(nice_TD):
```

Comment:
I also thought about doing it inside `make_nice_tree_decomposition` function directly, given the relatively small overhead. But I'm thinking about taking a modular approach, and don't want to make the nice tree decomp function too complicated (since this function or `treewidth` already have many arguments). So I am slightly unsure which approach to take.